test/interface/jacobian.jl

```diff
@@ -8,49 +8,180 @@ using Test
+    njacs[] += 1
     J[1, 1] = 1 - h(p, t - 1)[1]
     nothing
   end
 
+  nWfacts = Ref(0)
+  function Wfact(W, u, h, p, dtgamma, t)
+    nWfacts[] += 1
+    W[1,1] = dtgamma * (1 - h(p, t - 1)[1]) - 1
+    nothing
+  end
+
+  nWfact_ts = Ref(0)
+  function Wfact_t(W, u, h, p, dtgamma, t)
+    nWfact_ts[] += 1
+    W[1,1] = 1 - h(p, t - 1)[1] - inv(dtgamma)
+    nothing
+  end
+
   h(p, t) = [0.0]
 
   # define problems
-  prob_wo_jac = DDEProblem(DDEFunction{true}(f), [1.0], h, (0.0, 40.0);
-                           constant_lags = [1])
-  prob_w_jac = DDEProblem(DDEFunction{true}(f; jac = g), [1.0], h, (0.0, 40.0);
-                          constant_lags = [1])
+  prob = DDEProblem(DDEFunction{true}(f), [1.0], h, (0.0, 40.0); constant_lags = [1])
+  prob_jac = remake(prob; f = DDEFunction{true}(f; jac = jac))
+  prob_Wfact = remake(prob; f = DDEFunction{true}(f; Wfact = Wfact))
+  prob_Wfact_t = remake(prob; f = DDEFunction{true}(f; Wfact_t = Wfact_t))
 
   # compute solutions
   for alg in (Rosenbrock23(), TRBDF2())
-    sol_wo_jac = solve(prob_wo_jac, MethodOfSteps(alg))
-    sol_w_jac = solve(prob_w_jac, MethodOfSteps(alg))
+    sol = solve(prob, MethodOfSteps(alg))
+
+    ## Jacobian
+    njacs[] = 0
+    sol_jac = solve(prob_jac, MethodOfSteps(alg))
+    
+    # check number of function evaluations
+    @test !iszero(njacs[])
+    @test njacs[] == sol_jac.destats.njacs
+    if alg isa Rosenbrock23
+      @test njacs[] == sol_jac.destats.nw
+    else
+      @test_broken njacs[] == sol_jac.destats.nw
+    end
+
+    # check resulting solution
+    @test sol.t ≈ sol_jac.t
+    @test sol.u ≈ sol_jac.u
+
+    ## Wfact
+    nWfacts[] = 0
+    sol_Wfact = solve(prob_Wfact, MethodOfSteps(alg))
+
+    # check number of function evaluations
+    if alg isa Rosenbrock23
+      @test !iszero(nWfacts[])
+      @test nWfacts[] == njacs[]
+      @test iszero(sol_Wfact.destats.njacs)
+    else
+      @test_broken !iszero(nWfacts[])
+      @test_broken nWfacts[] == njacs[]
+      @test_broken iszero(sol_Wfact.destats.njacs)
+    end
+    @test_broken nWfacts[] == sol_Wfact.destats.nw
+
+    # check resulting solution
+    @test sol.t ≈ sol_Wfact.t
+    @test sol.u ≈ sol_Wfact.u
+
+    ## Wfact_t
+    nWfact_ts[] = 0
+    sol_Wfact_t = solve(prob_Wfact_t, MethodOfSteps(alg))
+
+    # check number of function evaluations
+    if alg isa Rosenbrock23
+      @test_broken !iszero(nWfact_ts[])
+      @test_broken nWfact_ts[] == njacs[]
+      @test_broken iszero(sol_Wfact_t.destats.njacs)
+    else
+      @test !iszero(nWfact_ts[])
+      @test_broken nWfact_ts[] == njacs[]
+      @test iszero(sol_Wfact_t.destats.njacs)
+    end
+    @test_broken nWfact_ts[] == sol_Wfact_t.destats.nw
 
-    @test sol_wo_jac.t ≈ sol_w_jac.t
-    @test sol_wo_jac.u ≈ sol_w_jac.u
+    # check resulting solution
+    if alg isa Rosenbrock23
+      @test sol.t ≈ sol_Wfact_t.t
+      @test sol.u ≈ sol_Wfact_t.u
+    else
+      @test_broken sol.t ≈ sol_Wfact_t.t
```

Comment:
that means it's not working, since IIRC at this point only Rosenbrock23 is using the untransformed version still?

Reply:
Yeah, as I said, maybe it's a bit too ambitious at this point. The main motivation of the tests is that:
1. If we specify the Jacobian, the number of Jacobian and W evaluations should be equal to the non-zero value of our counter, and the solution should be the same.
2. If we specify Wfact or Wfact_t, the number of Jacobian evaluations should be zero and the number of W evaluations should be equal to the non-zero value of our counter, and the solution should be the same.

Reply:
Oh, I see: the time points are equal, not that the value is correct. Yeah, I am not sure this will be the case, since symbolic CSE stuff happens on the symbolic factorization (the main source of the speedup)